distributed/rpc/batch/parameter_server.py

```diff
@@ -0,0 +1,144 @@
+
+            if self.curr_update_size >= self.batch_update_size:
+                for p in self.model.parameters():
+                    p.grad /= self.batch_update_size
+                self.curr_update_size = 0
+                self.optimizer.step()
+                self.optimizer.zero_grad()
+                fut.set_result(self.model)
+                timed_log("PS updated model")
+                self.future_model = torch.futures.Future()
+
+        return fut
+
+
+class Trainer(object):
+
+    def __init__(self, ps_rref):
+        self.ps_rref = ps_rref
+        self.loss_fn = nn.MSELoss()
+        self.one_hot_indices = torch.LongTensor(batch_size) \
+                                    .random_(0, num_classes) \
+                                    .view(batch_size, 1)
+
+    def get_next_batch(self):
+        for _ in range(6):
+            inputs = torch.randn(batch_size, 3, image_w, image_h)
+            labels = torch.zeros(batch_size, num_classes) \
+                        .scatter_(1, self.one_hot_indices, 1)
+            yield inputs.cuda(), labels.cuda()
+
+    def train(self):
+        name = rpc.get_worker_info().name
+        m = self.ps_rref.rpc_sync().get_model().cuda()
+        for inputs, labels in self.get_next_batch():
+            timed_log(f"{name} processing one batch")
+            self.loss_fn(m(inputs), labels).backward()
+            timed_log(f"{name} reporting grads")
+            m = rpc.rpc_sync(
+                self.ps_rref.owner(),
+                BatchUpdateParameterServer.update_and_fetch_model,
+                args=(self.ps_rref, [p.grad for p in m.cpu().parameters()]),
+            ).cuda()
+            timed_log(f"{name} got updated model")
+
+
+def run_trainer(ps_rref):
+    trainer = Trainer(ps_rref)
+    trainer.train()
+
+
+def run_ps(trainers):
+    timed_log("Start training")
+    ps_rref = rpc.RRef(BatchUpdateParameterServer())
+    futs = []
+    for trainer in trainers:
+        futs.append(
+            rpc.rpc_async(trainer, run_trainer, args=(ps_rref,))
+        )
+
+    torch.futures.wait_all(futs)
+    timed_log("Finish training")
+
+
+def run(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '29500'
+    options=rpc.ProcessGroupRpcBackendOptions(
+        num_send_recv_threads=16,
+        rpc_timeout=0  # infinite timeout
+     )
+    if rank != 0:
+        rpc.init_rpc(
+            f"trainer{rank}",
+            rank=rank,
+            world_size=world_size,
+            rpc_backend_options=options
+        )
+        # trainer passively waiting for ps to kick off training iterations
+    else:
+        rpc.init_rpc(
+            "ps",
+            rank=rank,
+            world_size=world_size,
+            rpc_backend_options=options
+        )
+        run_ps([f"trainer{r}" for r in range(1, world_size)])
+
+    # block until all rpcs finish
+    rpc.shutdown()
+
+
+if __name__=="__main__":
+    world_size = 6
```

Comment:
IIUC there needs to be at least `batch_update_size` trainers, otherwise, in the first iteration, the trainer will never get unblocked from the RPC that waits for 5 updates right? Should we code this value to `world_size = batch_update_size + 1` ?

Reply:
Good catch! Hard-coding these two args are bad. Let me update